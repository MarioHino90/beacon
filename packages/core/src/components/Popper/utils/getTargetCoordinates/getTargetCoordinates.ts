import { RefObject } from "react";

import { TargetCoordinates } from "../../types";

/**
 * Get coordinates for the target element on the DOM
 * @param element an HTML element instance
 * @param portalRef ref of Portal inside which Popper will be rendered
 */
export const getTargetCoordinates = <P extends HTMLElement = HTMLDivElement>(
  element: HTMLElement,
  portalRef?: RefObject<P>
): TargetCoordinates => {
  // calculate any offsets based on scroll and/or page layout behavior
  const pageLeftOffset =
    window.pageXOffset || document.documentElement.scrollLeft;
  const pageTopOffset =
    window.pageYOffset || document.documentElement.scrollTop;

  // get the coordinates of the target element
  const rect = element.getBoundingClientRect();

  // if portalRef.current is defined, then calculate element's
  // positioning relative to the portal parent
  if (portalRef?.current) {
    // get portalRef's parent bounding rect
    const parentRect = portalRef.current.getBoundingClientRect();

    // return relative coordinates
    return {
      top: rect.top - parentRect.top + pageTopOffset,
      right: rect.width,
      bottom: rect.height,
      left: rect.left - parentRect.left + pageLeftOffset,
    };
  }

  // if portalRef.current isn't defined,
  // then return the absolute coords
  return {
    top: rect.top + pageTopOffset,
    right: rect.width,
    bottom: rect.height,
    left: rect.left + pageLeftOffset,
  };
};
