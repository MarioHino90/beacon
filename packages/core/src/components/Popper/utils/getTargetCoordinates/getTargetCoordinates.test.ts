import { getTargetCoordinates } from "./getTargetCoordinates";

describe(`getTargetCoordinates`, () => {
  it(`should match snapshot`, () => {
    // create a fake popper
    const popper = document.createElement("div");
    const content = document.createTextNode("popper");

    popper.appendChild(content);

    // mount to DOM
    document.body.appendChild(popper);

    // make assertion
    expect(getTargetCoordinates(popper)).toMatchSnapshot();

    // remove from DOM as cleanup stage
    document.body.removeChild(popper);
  });

  it(`should match snapshot when 'portalRef' is defined`, () => {
    // create a fake popper
    const popper = document.createElement("div");
    const content = document.createTextNode("popper");

    popper.appendChild(content);

    // create a portal
    const portalParent = document.createElement("div");

    portalParent.appendChild(popper);

    // mount to DOM
    document.body.appendChild(portalParent);

    // make assertion
    expect(
      getTargetCoordinates(popper, { current: portalParent })
    ).toMatchSnapshot();

    // remove from DOM as cleanup stage
    document.body.removeChild(portalParent);
  });
});
