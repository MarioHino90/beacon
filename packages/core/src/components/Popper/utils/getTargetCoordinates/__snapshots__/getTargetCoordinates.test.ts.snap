// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTargetCoordinates should match snapshot 1`] = `
{
  "bottom": 0,
  "left": 0,
  "right": 0,
  "top": 0,
}
`;

exports[`getTargetCoordinates should match snapshot when 'portalRef' is defined 1`] = `
{
  "bottom": 0,
  "left": 0,
  "right": 0,
  "top": 0,
}
`;
